### PR TITLE
EZP-32155: Add a custom limitation example (3.3)

### PR DESCRIPTION
**Pull request by adriendupuis**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | [EZP-32155](https://issues.ibexa.co/browse/EZP-32155)
| Versions      | 3.3

* Custom Limitation example
* Custom Policy adjustment

To use examples:
* ```shell
  cp -rv $DOC_DIR/code_samples/back_office/limitation/* $DXP_DIR/
  cat $DXP_DIR/config/append_to_services.yaml >> $DXP_DIR/config/services.yaml
  rm $DXP_DIR/config/append_to_services.yaml
  ```
* Edit a random Role like Anonymous
* Add/Create a `Custom module / Function #2` policy
* Check or not the `Custom limitation`
* Update the policy

### CODE REVIEW

File: docs/guide/custom_policies.md [hunk 1]
```diff
@@ -119,40 +119,168 @@ class Kernel extends BaseKernel
...
 ```yaml
 App\Security\Limitation\Mapper\CustomLimitationValueMapper:
     arguments:
         # ...
     tags:
-        - { name: 'ez.limitation.valueMapper', limitationType: 'Custom' }
+        - { name: 'ez.limitation.valueMapper', limitationType: 'CustomLimitation' }
 ```
 
-If you want to completely override the way of rendering custom Limitation values in the role view,
```
Comment (adriendupuis) [inline docs/guide/custom_policies.md:148]: This is not optional but mandatory

`An exception has been thrown during the rendering of a template ("Could not find a block for CustomLimitation: ez_limitation_customlimitation_value.").`

File: docs/guide/custom_policies.md [hunk 2]
```diff
@@ -163,22 +291,76 @@ ezplatform:
...
-
+                - { template: '@ezdesign/limitation/custom_limitation_value.html.twig', priority: 0 }
 ```
 
-!!! note
+To check if current user has this custom limitation set to true from a custom controller:
+```php
+<?php
 
-    If you skip this part, Limitation values will be rendered using an [`ez_limitation_value_fallback`](https://github.com/ezsystems/repository-forms/blob/master/bundle/Resources/views/limitation_values.html.twig#L1-L6) block as comma-separated list.
```
Comment (adriendupuis) [inline docs/guide/custom_policies.md:172]: Have a better look to this. Maybe useful when using `MultipleSelectionBasedMapper`.

Comment (adriendupuis) [reply docs/guide/custom_policies.md:172]: https://github.com/ezsystems/repository-forms/blob/2.5/bundle/Resources/views/limitation_values.html.twig#L1
TODO: try if this fallback is used when there is no CustomLimitationValueMapper service.

Comment (adriendupuis) [reply docs/guide/custom_policies.md:176]: The fallback is used when the CustomLimitationValueMapper service is commented out.

File: docs/guide/custom_policies.md (hunk 1 above)
Comment (adriendupuis) [reply docs/guide/custom_policies.md:152]: As soon as a value mapper exist, the template is mandatory.

File: docs/guide/custom_policies.md [hunk 3]
```diff
@@ -6,14 +6,14 @@ description: Create a custom Policy to cover non-standard permission needs.
 
 The content Repository uses [Roles and Policies](permissions.md) to give Users access to different functions of the system.
 
-Any bundle can expose available Policies via a `PolicyProvider` which can be added to EzPublishCoreBundle's [service container](../api/public_php_api.md#service-container) extension.
+Any bundle can expose available Policies via a `PolicyProvider` which can be added to EzPublishCoreBundle's [service container](../api/public_php_api.md#service-container) extension.
```
Comment (adriendupuis) [inline docs/guide/custom_policies.md:9]:
```suggestion
Any bundle can expose available Policies via a `PolicyProvider` which can be added to EzPublishCoreBundle's [service container](public_php_api.md#service-container) extension.
```

File: docs/guide/custom_policies.md [hunk 4]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
 
-## Integrating custom Limitation types with the UI
+### Custom Limitation type form
+
+#### Form mapper
 
 To provide support for editing custom policies in the Back Office, you need to implement [`EzSystems\EzPlatformAdminUi\Limitation\LimitationFormMapperInterface`](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/src/lib/Limitation/LimitationFormMapperInterface.php).
 
+- `mapLimitationForm` adds the limitation field as a child to a provided Symfony form.
+- `getFormTemplate` returns the path to the template to use for rendering the limitation form. Here it use [`form_label`](https://symfony.com/doc/5.4/form/form_customization.html#reference-forms-twig-label) and [`form_widget`](https://symfony.com/doc/5.4/form/form_customization.html#reference-forms-twig-widget) to do so.
```
Comment (adriendupuis) [inline docs/guide/custom_policies.md:141]:
```suggestion
- `getFormTemplate` returns the path to the template to use for rendering the limitation form. Here it use [`form_label`]([[= symfony_doc =]]/form/form_customization.html#reference-forms-twig-label) and [`form_widget`]([[= symfony_doc =]]/form/form_customization.html#reference-forms-twig-widget) to do so.
```

File: docs/guide/custom_policies.md [hunk 5]
```diff
@@ -6,14 +6,14 @@ description: Create a custom Policy to cover non-standard permission needs.
...
 
 ## PolicyProvider
 
-A `PolicyProvider` object provides a hash containing declared modules, functions and Limitations.
+A `PolicyProvider` object provides a hash containing declared modules, functions and Limitations.
 
-- Each Policy provider provides a collection of permission *modules*.
-- Each module can provide *functions* (e.g. in `content/read` "content" is the module, "read" is the function)
+- Each Policy provider provides a collection of permission *modules*.
+- Each module can provide *functions* (For example, in `content/read`, "content" is the module, and "read" is the function)
```
Comment (adriendupuis) [inline docs/guide/custom_policies.md:16]:
```suggestion
- Each module can provide *functions* (for example, in `content/read`, "content" is the module, and "read" is the function)
```

File: docs/guide/custom_policies.md [hunk 6]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+[[= include_file('code_samples/back_office/limitation/translations/ezplatform_content_forms_policies.en.yaml') =]]
 ```
 
-!!! note
+### Checking user custom Limitation
 
-    If you skip this part, Limitation values will be rendered using an [`ez_limitation_value_fallback`](https://github.com/ezsystems/repository-forms/blob/master/bundle/Resources/views/limitation_values.html.twig#L1-L6) block as comma-separated list.
+To check if current user has this custom limitation set to true from a custom controller:
+```php
+<?php
```
Comment (adriendupuis) [inline docs/guide/custom_policies.md:212]:
```suggestion
<?php declare(strict_types=1);
```

File: docs/guide/custom_policies.md [hunk 7]
```diff
@@ -33,12 +33,14 @@ by the alias declared in `LimitationType` service tag. If no Limitation is provi
...
 ```
 
-Limitations need to be implemented as *Limitation types* and declared as services identified with `ezpublish.limitationType` tag.
+Limitations need to be implemented as *Limitation types* and declared as services identified with `ezpublish.limitationType` tag.
 Name provided in the hash for each Limitation is the same value set in the `alias` attribute in the service tag.
 
 For example:
 
 ``` php
+<?php
```
Comment (adriendupuis) [inline docs/guide/custom_policies.md:42]:
```suggestion
<?php declare(strict_types=1);
```

File: docs/guide/custom_policies.md [hunk 8]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
-    tags:
-        - { name: 'ez.limitation.formMapper', limitationType: 'Custom' }
+``` yaml
+[[= include_file('code_samples/back_office/limitation/config/append_to_services.yaml', 5, 8) =]]
 ```
 
-If you want to provide human-readable names of the custom Limitation values, you need to implement [`EzSystems\EzPlatformAdminUi\Limitation\LimitationValueMapperInterface`](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/src/lib/Limitation/LimitationValueMapperInterface.php).
+#### Notable form mappers to extend
+
+Some abstract Limitation type form mapper classes are provided to help implementing common complexe Limitations.
```
Comment (adriendupuis) [inline docs/guide/custom_policies.md:162]:
```suggestion
Some abstract Limitation type form mapper classes are provided to help implementing common complex Limitations.
```

**Comment by adriendupuis** [comment]
TODO: Update #2195, the equivalent for 4.5+, after review and merge.

**Review by adamwojs — APPROVED** [review]
+1 by please remember about code re-branding during merge up

**Comment by adriendupuis** [comment]
> +1 by please remember about code re-branding during merge up

@adamwojs I'm pretty sure I forgot nothing in #2195 but will send it to review for being sure 😉

File: code_samples/back_office/limitation/translations/forms.en.yaml [hunk 9]
```diff
@@ -0,0 +1,4 @@
+role.policy.custom_module: 'Custom module'
+role.policy.custom_module.all_functions: 'Custom module / All functions'
+role.policy.custom_module.custom_function_1: 'Custom module / Function #1'
+role.policy.custom_module.custom_function_2: 'Custom module / Function #2'
```
Comment (adriendupuis) [inline code_samples/back_office/limitation/translations/forms.en.yaml:4]: TODO: Study #2205 for an example allowing `translation:extract` to generate this file (something like `php bin/console translation:extract en --domain=forms --dir=src --output-dir=translations`)

File: docs/guide/custom_policies.md [hunk 10]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
-        
-        // Retrieve "ezpublish" container extension
-        $eZExtension = $container->getExtension('ezpublish');
-        // Add the policy provider
-        $eZExtension->addPolicyProvider(new MyPolicyProvider());
-    }
-}
+The base of a custom limitation is a class to store values for the usage of this limitation in roles, and a class to implement the limitation's logic.
+
+The value class extends `eZ\Publish\API\Repository\Values\User\Limitation` and says for which limitation it's used:
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:111]:
```suggestion
The value class extends `eZ\Publish\API\Repository\Values\User\Limitation` and specifies the limitation for which it's used:
```

File: docs/guide/custom_policies.md [hunk 11]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+
+``` php
+[[= include_file('code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationValue.php') =]]
+```
+
+The type class implements `eZ\Publish\SPI\Limitation\Type`.
+
+- `accept`, `validate` and `buildValue` implement the value class usage logic.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:119]:
```suggestion
- `accept`, `validate` and `buildValue` implement the logic for using the value class.
```

File: docs/guide/custom_policies.md [hunk 12]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+- `evaluate` challenges a limitation value against the current user, the subject object and other context objects to return if the limitation is satisfied or not. `evaluate` is, among others, used by `PermissionResolver::canUser` (to check if a user having access to a function can use it in its limitations) and `PermissionResolver::lookupLimitations`.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:120]:
```suggestion
- `evaluate` assesses a limitation value against the current user, the subject object, and other context objects to determine if the limitation is satisfied or not. `evaluate` is used, among others places, by `PermissionResolver::canUser` (to check if a user with access to a function can use it in its limitations) and `PermissionResolver::lookupLimitations`.
```

File: docs/guide/custom_policies.md [hunk 13]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+
+```php
+[[= include_file('code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php') =]]
+```
+
+The type class is set as a service tagged `ezpublish.limitationType` with an alias to identify it, and to link it to the value.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:126]:
```suggestion
The type class is registered as a service tagged with `ezpublish.limitationType` and given an alias to identify it, as well as to link it to the value.
```

File: docs/guide/custom_policies.md [hunk 14]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
 
-## Integrating custom Limitation types with the UI
+### Custom Limitation type form
+
+#### Form mapper
 
 To provide support for editing custom policies in the Back Office, you need to implement [`EzSystems\EzPlatformAdminUi\Limitation\LimitationFormMapperInterface`](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/src/lib/Limitation/LimitationFormMapperInterface.php).
 
+- `mapLimitationForm` adds the limitation field as a child to a provided Symfony form.
+- `getFormTemplate` returns the path to the template to use for rendering the limitation form. Here it use [`form_label`]([[= symfony_doc =]]/form/form_customization.html#reference-forms-twig-label) and [`form_widget`]([[= symfony_doc =]]/form/form_customization.html#reference-forms-twig-widget) to do so.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:141]:
```suggestion
- `getFormTemplate` returns the path to the template to use for rendering the limitation form. Here it uses [`form_label`]([[= symfony_doc =]]/form/form_customization.html#reference-forms-twig-label) and [`form_widget`]([[= symfony_doc =]]/form/form_customization.html#reference-forms-twig-widget) to do so.
```

File: docs/guide/custom_policies.md [hunk 15]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+``` yaml
+[[= include_file('code_samples/back_office/limitation/config/append_to_services.yaml', 5, 8) =]]
 ```
 
-If you want to provide human-readable names of the custom Limitation values, you need to implement [`EzSystems\EzPlatformAdminUi\Limitation\LimitationValueMapperInterface`](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/src/lib/Limitation/LimitationValueMapperInterface.php).
+#### Notable form mappers to extend
+
+Some abstract Limitation type form mapper classes are provided to help implementing common complex Limitations.
+
+- `MultipleSelectionBasedMapper` is mapper to build form for Limitation based on checkbox list where multiple items can be chosen. For example, it's used to build forms for [Content Type Limitation](limitation_reference.md#content-type-limitation), [Language Limitation](limitation_reference.md#language-limitation) or [Section Limitation](limitation_reference.md#section-limitation).
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:164]:
```suggestion
- `MultipleSelectionBasedMapper` is a mapper used to build forms for Limitation based on checkbox lists, where multiple items can be chosen. For example, it's used to build forms for [Content Type Limitation](limitation_reference.md#content-type-limitation), [Language Limitation](limitation_reference.md#language-limitation) or [Section Limitation](limitation_reference.md#section-limitation).
```

File: docs/guide/custom_policies.md [hunk 16]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+- `UDWBasedMapper` is to build Limitation form where a Content/Location must be selected. For example, it's used by the [Subtree Limitation](limitation_reference.md#subtree-of-location-limitation) form.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:165]:
```suggestion
- `UDWBasedMapper` is used to build Limitation form where a Content/Location must be selected. For example, it's used by the [Subtree Limitation](limitation_reference.md#subtree-of-location-limitation) form.
```

File: docs/guide/custom_policies.md [hunk 17]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+
+#### Value mapper
+
+By default, without a value mapper, the Limitation value is rendered using the block `ez_limitation_value_fallback` of the template [`vendor/ezsystems/ezplatform-admin-ui/src/bundle/Resources/views/themes/admin/limitation/limitation_values.html.twig`](https://github.com/ezsystems/ezplatform-admin-ui/blob/2.3/src/bundle/Resources/views/themes/admin/limitation/limitation_values.html.twig#L1-L6).
+
+To customize the rendering, a value mapper eventually transforms the Limitation value and send it to a custom template.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:171]:
```suggestion
To customize the rendering, a value mapper eventually transforms the Limitation value and sends it to a custom template.
```

File: docs/guide/custom_policies.md [hunk 18]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+
+The value mapper implements [`EzSystems\EzPlatformAdminUi\Limitation\LimitationValueMapperInterface`](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/src/lib/Limitation/LimitationValueMapperInterface.php).
+
+Its `mapLimitationValue` function returns the Limitation value transformed for the needs of the template.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:175]:
```suggestion
Its `mapLimitationValue` function returns the Limitation value adjusted to the needs of the template.
```
OR
```suggestion
Its `mapLimitationValue` function returns the Limitation value transformed to meet the requirements of the the template.
```

File: docs/guide/custom_policies.md [hunk 19]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
-    tags:
-        - { name: 'ez.limitation.valueMapper', limitationType: 'Custom' }
+``` yaml
+[[= include_file('code_samples/back_office/limitation/config/append_to_services.yaml', 9, 12) =]]
 ```
 
-If you want to completely override the way of rendering custom Limitation values in the role view,
-create a Twig template containing block definition which follows the naming convention:
-`ez_limitation_<LIMITATION TYPE>_value`. For example:
+When a value mapper exists for a Limitation, the rendering uses a Twig block named `ez_limitation_<lower_case_identifier>_value` where `<lower_case_identifier>` is the Limitation identifier in lower case.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:187]:
```suggestion
When a value mapper exists for a Limitation, the rendering uses a Twig block named `ez_limitation_<lower_case_identifier>_value` where `<lower_case_identifier>` is the Limitation identifier in a lower case.
```

File: docs/guide/custom_policies.md [hunk 20]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+In this example, block name is `ez_limitation_customlimitation_value` as the identifier is `CustomLimitation`.
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:188]:
```suggestion
In this example, `ez_limitation_customlimitation_value` is a block name and  `CustomLimitation` is an identifier.
```

File: docs/guide/custom_policies.md [hunk 21]
```diff
@@ -58,131 +60,203 @@ class MyPolicyProvider implements PolicyProviderInterface
...
 
+``` yaml
+[[= include_file('code_samples/back_office/limitation/translations/ezplatform_content_forms_policies.en.yaml') =]]
 ```
 
-!!! note
+### Checking user custom Limitation
 
-    If you skip this part, Limitation values will be rendered using an [`ez_limitation_value_fallback`](https://github.com/ezsystems/repository-forms/blob/master/bundle/Resources/views/limitation_values.html.twig#L1-L6) block as comma-separated list.
+To check if current user has this custom limitation set to true from a custom controller:
```
Comment (MagdalenaZuba) [inline docs/guide/custom_policies.md:210]:
```suggestion
To check if the current user has this custom limitation set to true from a custom controller:

```

**Review by MagdalenaZuba — CHANGES_REQUESTED** [review]

File: docs/guide/custom_policies.md (hunk 20 above)
Comment (adriendupuis) [reply docs/guide/custom_policies.md:188]: @MagdalenaZuba 
```suggestion
In this example, `ez_limitation_customlimitation_value` is the block name, and `CustomLimitation` is the identifier.
```

**Review by MagdalenaZuba — APPROVED** [review]

**Pull request merged**